datafusion/core/src/physical_optimizer/join_selection.rs

```diff
@@ -108,197 +105,49 @@ fn supports_collect_by_thresholds(
-    right_schema_len: usize,
-    projection: Option<&Vec<usize>>,
-    join_type: &JoinType,
-) -> Option<Vec<usize>> {
-    match join_type {
-        // For Anti/Semi join types, projection should remain unmodified,
-        // since these joins output schema remains the same after swap
-        JoinType::LeftAnti
-        | JoinType::LeftSemi
-        | JoinType::RightAnti
-        | JoinType::RightSemi => projection.cloned(),
-
-        _ => projection.map(|p| {
-            p.iter()
-                .map(|i| {
-                    // If the index is less than the left schema length, it is from
-                    // the left schema, so we add the right schema length to it.
-                    // Otherwise, it is from the right schema, so we subtract the left
-                    // schema length from it.
-                    if *i < left_schema_len {
-                        *i + right_schema_len
-                    } else {
-                        *i - left_schema_len
-                    }
-                })
-                .collect()
-        }),
-    }
+    join_type.swap()
 }
 
 /// This function swaps the inputs of the given join operator.
 /// This function is public so other downstream projects can use it
 /// to construct `HashJoinExec` with right side as the build side.
+#[deprecated(since = "45.0.0", note = "use HashJoinExec::swap_inputs instead")]
 pub fn swap_hash_join(
     hash_join: &HashJoinExec,
     partition_mode: PartitionMode,
 ) -> Result<Arc<dyn ExecutionPlan>> {
-    let left = hash_join.left();
-    let right = hash_join.right();
-    let new_join = HashJoinExec::try_new(
-        Arc::clone(right),
-        Arc::clone(left),
-        hash_join
-            .on()
-            .iter()
-            .map(|(l, r)| (Arc::clone(r), Arc::clone(l)))
-            .collect(),
-        swap_join_filter(hash_join.filter()),
-        &swap_join_type(*hash_join.join_type()),
-        swap_join_projection(
-            left.schema().fields().len(),
-            right.schema().fields().len(),
-            hash_join.projection.as_ref(),
-            hash_join.join_type(),
-        ),
-        partition_mode,
-        hash_join.null_equals_null(),
-    )?;
-    // In case of anti / semi joins or if there is embedded projection in HashJoinExec, output column order is preserved, no need to add projection again
-    if matches!(
-        hash_join.join_type(),
-        JoinType::LeftSemi
-            | JoinType::RightSemi
-            | JoinType::LeftAnti
-            | JoinType::RightAnti
-    ) || hash_join.projection.is_some()
-    {
-        Ok(Arc::new(new_join))
-    } else {
-        // TODO avoid adding ProjectionExec again and again, only adding Final Projection
-        let proj = ProjectionExec::try_new(
-            swap_reverting_projection(&left.schema(), &right.schema()),
-            Arc::new(new_join),
-        )?;
-        Ok(Arc::new(proj))
-    }
+    hash_join.swap_inputs(partition_mode)
```

Comment:
This is the key point of this PR -- to move this (and related) function into the HashJoinExec itself